### PR TITLE
Bring OCR to Mealie for importing scanned recipes

### PR DESCRIPTION
**Pull request by Miroito**
# Added so far
- New tab in the recipe creation page for scanned recipes (I'm open for the icon that should be used)
- ~~As a first draft, the recipe is created with the recognized text in the description field to copy and paste later in edit mode.~~ no more
**2022-07-28 Update**: I'm at a point where I think the component is very usable and open for reviews to merge something that is very close to what is already implemented. 

# Before merging
_Here the list of tasks before we consider acceptable to merge this code in the beta_

- [x] Add the experimental feature flag
- [x] Write tests for each service
- - [x] image_to_string (skipped because tesseract's output is not reliable between distros)
- - [x] image_to_tsv (skipped because tesseract's output is not reliable between distros)
- - [x] format_tsv_output
- [x] Restrict the possible file types
- [x] Checkbox make the uploaded picture the recipe thumbnail
- [x] Nice UI to make the process of importing a recipe like this easier and usable
- [x] Add possibility to automatically populate single ingredients/steps fields
- [x] Replace the POST tsv route to take asset names in order instead of files that are already on the server
- [x] Add a way to return to the ocr-editor page after a recipe was created
- [x] Remove sample ingredients and steps from the initial recipe creation
- [x] Tidy up the ui (looking especially at the buttons)
- [x] Preserve lines or paragraphs
- [x] Fix the things that makes the CI so angry
- [x] Write help dialog
- [x] Bugfix: Mouse position is offset when the page is scrolled down
- [x] Make all the hard coded English text translatable
- [x] ~~Add mode to highlight boxes of recognized text on the image~~ Recognized text is highlighted on component mount
- [x] Fix bug where ratio is not respected on big images
- [ ] Add a quick(er) help for users to get going as fast as possible. 

# Nice to have's
- [x] Define the canvas variable pointing to the canvas html element only once instead of every function in the ocr-editor component
- [ ] Add advanced settings (e.g. Language to improve pytesseract recognition)
- [ ] Possibility to add multiple pages/files and switch between them
- [x] Clean up the ocr-editor page by creating a RecipeOcrEditor component
- [ ] Automatic field filling suggestion
- - [x] Recipe title

# Design
This new feature is based on previous experience with a similar software solution called Esker.
The process that I have designed for now lets the user use a new creation page` /recipe/create/ocr` letting them upload a picture, optionally making it the recipe thumbnail. This creates a recipe called "New OCR Recipe" with the uploaded picture as an asset called "Original recipe image". Additionally, a new column in the recipes table registers that this recipe is an OCR recipe. 

The user is directed to the page "recipe/_slug/ocr-editor" where they can use the image they uploaded to fill the usual recipe fields on the right part of the page. When this page in mounted, it sends the asset name to the backend fot it to send back the text and contained inside and its position.

Two modes are available.
- Selection mode, lets the user input data.
- Pan and Zoom mode lets the user move around when pictures are big enough to do so.

In selection mode, the user can draw a rectangle, the identified text will appear under the canvas. The user can then select any recipe field on the right, then click anywhere inside the rectangle. This will take whatever text is fully contained in the rectangle and overwrite the field that was last selected.

The bulk add buttons will spawn a dialog with the selected text (understand text under the drawn rectangle) inside them.
This is where the Split text modes come into play, it lets the user choose whether they want to keep all line breaks, for example, if a recipe book  lists one ingredient per line, they are able to select the whole list, press bulk add on the ingredient tab and add all ingredients in 2 clicks.

The mode flatten will remove all line breaks and the blocks mode will put line breaks between identified blocks by tesseract. The blocks mode is pretty useful for instructions, that usually come into multiple paragraphs in a form of blocks, making it easier to use the bulk add dialog, this time for instructions.

For recipes that are called `New OCR Recipe (n)` or regex `/New\sOCR\sRecipe(\s\([0-9]+\))?/g`, the ocr-editor component will take the biggest block with the fewer words, assume it is the recipe's title, and populate it in the recipe name field. This is done with the function `findRecipeTitle` in the ocr-editor component.

When the user is happy with the edits the recipe can be saved the usual way.They can come back to the OCR editor page by clicking the usual edit button and using the new button "OCR Editor" that will appear when the recipe is an OCR Recipe (hence the new table column).

### CODE REVIEW

**Comment by hay-kot** [comment]
FYI Rebasing should get CI sorted. 

Need the changes from #1252

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ repos:
       - id: check-toml
       - id: end-of-file-fixer
       - id: trailing-whitespace
+        exclude-types: ["txt","tsv"]
```
Comment (Miroito) [inline .pre-commit-config.yaml]: I had to add this line to prevent the pre-commit hook to trim whitespaces at the end of lines on this expected result file.
If I the end-of-line whitespaces are removed, the test fails.

**Comment by hay-kot** [comment]
Problem with your CI was your poetry lock file

CleanShot 2022-07-28 at 14 29 47@2x

Now it's failing for other reasons 😄

**Comment by Miroito** [comment]
Thanks I missed this line in the feedback.

Yes sounds like the type check is pretty angry at me for being sloppy. I'll get it to work.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,6 +9,7 @@ repos:
       - id: check-toml
       - id: end-of-file-fixer
       - id: trailing-whitespace
+        exclude: ^tests/
```
Comment (hay-kot) [inline .pre-commit-config.yaml]: Why are we excluding the entire tests directory?

File: frontend/api/class-interfaces/ocr.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    const formData = new FormData();
+    formData.append("file", file);
+    return await this.requests.post(`${prefix}/ocr/file-to-tsv`, formData);
+  }
+
+  async asset_to_tsv(recipeSlug: string, assetName: string) {
+    return await this.requests.post(`${prefix}/ocr/asset-to-tsv`, {recipeSlug, assetName});
+  }
+
+}
```
Comment (hay-kot) [inline frontend/api/class-interfaces/ocr.ts]: Can you camel case the names to conform to existing style?

File: mealie/schema/recipe/recipe.py [hunk 4]
```diff
@@ -141,10 +141,11 @@ class Recipe(RecipeSummary):
     nutrition: Optional[Nutrition]
 
     # Mealie Specific
-    settings: Optional[RecipeSettings] = RecipeSettings()
+    settings: Optional[RecipeSettings] = None
```
Comment (hay-kot) [inline mealie/schema/recipe/recipe.py:144]: Why was this change required for OCR?

**Review by hay-kot — COMMENTED** [review]
These are just my cursory review comments. I haven't done a thorough review and there will likely be more changes required to get this merged in. 

Before I dig too deep into this I would like to see a more thorough write-up of the feature overall, how it works. 

Some critical areas I see that need more documentation are `frontend/pages/recipe/_slug/ocr-editor.vue` and some context around the tests that you've written like what is being tested and how we can validate the skipped test if the output is some-what unreliable. 

Looks really cool so far, thanks for your work on this one!

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Miroito) [reply .pre-commit-config.yaml]: This was forgotten, I meant to exclude tests/data because the tsv output has trailing whitespaces that pre-commit wanted to remove, making the test fail.
These tests are anyway unfortunately skipped at the moment, I kept them as skipped because I still want to use them, probably with less strict assertions. I'm still thinking about the best way to loosen up these tests so they pass.

File: frontend/api/class-interfaces/ocr.ts (hunk 3 above)
Comment (Miroito) [reply frontend/api/class-interfaces/ocr.ts]: Done, good find.

File: mealie/schema/recipe/recipe.py (hunk 4 above)
Comment (Miroito) [reply mealie/schema/recipe/recipe.py:144]: In short, giving the settings property of a Recipe a default value through RecipeSettings() meant it was not possible to create a recipe with custom settings given that I want that recipes created by OCR show assets by default.
This change is part of commit [90e1b85](https://github.com/hay-kot/mealie/pull/1244/commits/90e1b85135610b3364f5e2c190869123f5cbd042) where you will see that I also modified the `create_one` function of the recipe service to check if recipe settings are set, if not set them to the group's default.
Before that, when you passed a Recipe to this function, it would overwrite the settings with the group's defaults regardless if they were set or not.
To make it nice then, I set the default for the settings property to `None`, making sure that when a whole Recipe, we know whether someone put custom settings or whether we should use the group's defaults as defaults.

This change is part of this PR because I wanted to show assets by default on the recipes that are created using the OCR creation form. You will see in the diff of the above-mentioned commit that before making that change, I had to get the just created recipe via slug, set the custom setting and update the Recipe, which is less than sub-optimal.

**Comment by Miroito** [comment]
Added a design section to explain all the relevant work that make this possible hopefully the block of text is not too hard to read.
I want to add a limitations section later on to explain my thoughts on the current implementation and what can be done to make it more stable, reliable and functionally richer.

**Comment by Miroito** [comment]
Rebased to fix merge conflicts with mealie-next branch

**Comment by Miroito** [comment]
I have tried using the feature to add a bunch of recipes to see how it does. And... It's outputting gibberish when jpg files include a lot of text which is the worse user experience imaginable. There is either a lot a pre-processing to do that would make it much slower than it is currently.
Or, at least temporarily, I could either restrict the files to png or convert all images to png when they are uploded.

A little bit disappointed with tesseract, I'm going to invesitgate further as to why it is behaving this way,  

Putting the PR temporarily in draft again.

**Comment by Miroito** [comment]
I can't keep rebasing this amount of changes everytime so I have marked the PR up for review. Feedback would be greatly appreciated so we can merge this as soon as possible.

**Comment by hay-kot** [comment]
> I have tried using the feature to add a bunch of recipes to see how it does. And... It's outputting gibberish when jpg files include a lot of text which is the worse user experience imaginable. There is either a lot a pre-processing to do that would make it much slower than it is currently.
Or, at least temporarily, I could either restrict the files to png or convert all images to png when they are uploded.
>
> A little bit disappointed with tesseract, I'm going to invesitgate further as to why it is behaving this way,
>
> Putting the PR temporarily in draft again.

Any updates on this comment?

**Comment by Miroito** [comment]
> > I have tried using the feature to add a bunch of recipes to see how it does. And... It's outputting gibberish when jpg files include a lot of text which is the worse user experience imaginable. There is either a lot a pre-processing to do that would make it much slower than it is currently.
> > Or, at least temporarily, I could either restrict the files to png or convert all images to png when they are uploded.
> > A little bit disappointed with tesseract, I'm going to invesitgate further as to why it is behaving this way,
> > Putting the PR temporarily in draft again.
> 
> Any updates on this comment?

For now, I have restricted the image format to png, I have found no help on Tesseract's side though I did not look too far. 

There is also the option of converting any input image to png, then using the png to do the ocr. This adds a huge overhead that would mean I could not afford to ask the server every time to recognize characters in the image when the ocr editor is loaded like it is now.
I think in the long term, it is worth taking a little time to create the recipe and storing every ocr info the database. Creating the recipe the first time will take more time, but overall the recognition can probably be improved by preprocessing and the long processing would happen only once.
You might ask why I did not start by implementing it this way first; it seemed easier to do it the current way, to bring the feature "fast" to the main branch, the rest of the work can be done later as the current implementation is not blocking further improvements of the above nature and this is already a good base. 
I used it already a **few** times these past weeks by taking pictures of my books, creating the recipe on my dev instance and downloading the zip file to upload it to my usual mealie instance, I think it is already more than usable, improvements  can come later with user feedback as well.

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 5]
```diff
@@ -0,0 +1,922 @@
...
+          if (
+            state.selectedTextSplitMode !== "flatten" &&
+            index !== array.length - 1 &&
+            element[state.selectedTextSplitMode] !== array[index + 1][state.selectedTextSplitMode]
+          ) {
+            separator = "\n";
+          }
+          return element.text + separator;
+        })
+        .join("");
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: Can you provide some context on what's going on here through a code comment?

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 6]
```diff
@@ -0,0 +1,922 @@
...
+
+      // Don't let images bigger than the canvas be zoomed in more than 1:1 scale
+      // Meaning only let images smaller than the canvas to have a scale > 1
+      if (!state.isImageSmallerThanCanvas && state.imagePosition.scale > 1) {
+        state.imagePosition.scale = 1;
+      }
+    }
+
+    function findRecipeTitle() {
+      // This function will find the title of a recipe with the assumption that the title has the biggest ratio of surface area on number of words on the image
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: Use JSDoc style comments for function doc strings

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+          <p>
+            {{ $t("ocr-editor.help-dialog.split-modes.block-mode-desc") }}
+          </p>
+          <h2 class="my-2">
+            <v-icon> {{ $globals.icons.flatten }} </v-icon> {{ $t("ocr-editor.help-dialog.split-modes.flat-mode") }}
+          </h2>
+          <p>{{ $t("ocr-editor.help-dialog.split-modes.flat-mode-desc") }}</p>
+        </v-card-text>
+      </v-card>
+    </v-dialog>
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: Can we just show this at the bottom or top of the page? I think that would prevent a lot of Github Issues

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 8]
```diff
@@ -0,0 +1,922 @@
...
+                    v-bind="attrs"
+                    @click="switchSplitTextMode('flatten')"
+                    v-on="on"
+                  >
+                    {{ $globals.icons.flatten }}
+                  </v-icon>
+                </v-btn>
+              </template>
+              <span>{{ $t("ocr-editor.flatten") }}</span>
+            </v-tooltip>
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: These should really be rendered in some kind of look with a generic object type

File: frontend/pages/recipe/create.vue [hunk 9]
```diff
@@ -42,6 +42,11 @@ export default defineComponent({
         text: "Import with URL",
         value: "url",
       },
+      {
+        icon: $globals.icons.fileImage,
+        text: "Create recipe from an image",
```
Comment (hay-kot) [inline frontend/pages/recipe/create.vue]: Should be title cased like the rest of the menu, can you move this below "Import With .zip"

**Review by hay-kot — CHANGES_REQUESTED** [review]
Overall I think it looks pretty solid. There's a few small things that need to get cleanup up.

I'm not sure that I totally get the use case for this inside the application in it's current form. From what I experienced it only dumps the text you select in the pane below it and then the user copies and pastes it into the field? That workflow seems nearly identical to what they could already do with tooling on Windows and MacOs (probably linux?). There's definitely ways to improve this with thinks like auto-copy to clipboard or select a field and then highlight the corresponding text. 

That said I'm okay merging if you cleanup the items I commented on with the caveat that I may remove it later down the line if no one is interested in improving or maintaining it to make it more useful than existing tooling. 

Also if any conflicts come up before it gets merged I'll rebase and clean those up. Thanks for all your work on this one!

**Comment by Miroito** [comment]
> I'm not sure that I totally get the use case for this inside the application in it's current form. From what I experienced it only dumps the text you select in the pane below it and then the user copies and pastes it into the field? That workflow seems nearly identical to what they could already do with tooling on Windows and MacOs (probably linux?). There's definitely ways to improve this with thinks like auto-copy to clipboard or select a field and then highlight the corresponding text.

This is not right, but it shows the current design is not ready to be rolled out to unknowing people. I've written a help dialog that you can check out. In brief, select text on the image, click a field on the right, click on the rectangle. The text inside the rectangle gets propagated to the last selected recipe field.

> That said I'm okay merging if you cleanup the items I commented on with the caveat that I may remove it later down the line if no one is interested in improving or maintaining it to make it more useful than existing tooling.

It was a nice journey to write this app, you can say that you don't recognize the use case and close the PR, after all, I did write this without asking first. 
In my opinion, some people will definitely use this, given that they can easily find how to use it which does not seem to be the case just now. I mean, Mealie is amazing for recipe management, and I would love to be able to add some recipes from my books in around 1-2 minutes instead of 5-10 right now (given you use google lens and copy information for example).

> Also if any conflicts come up before it gets merged I'll rebase and clean those up. Thanks for all your work on this one!

Don't worry about it, I just wanted to start a review dialog so progress can be made. But yeah, rebasing alembic migrations is pure pain :)

**Comment by hay-kot** [comment]
I _do_ want to merge this as I think it can be super useful, I think it would probably be best to get an MVP merged and we can update the banner and link to a discussion or issues where people can provide direct feedback on usability and the interface. Then we can get some actionable feedback and make some improvements from there. 👍 

>  click a field on the right, click on the rectangle. The text inside the rectangle gets propagated to the last selected recipe field.

That's sort of how I expected it work but couldn't get it to work right off the bat. Entirely possible I just wasn't doing something right...

**Comment by Miroito** [comment]
> That's sort of how I expected it work but couldn't get it to work right off the bat. Entirely possible I just wasn't doing something right...

Yeah, I need to write a quick help on top of the page to explain in the fewest words possible how to get going fast, telling to go the help dialog for more details. I add this to the tasks on top.
I'm going to make the changes you highlighted as soon as I can, I switched to Endeavouros and somehow nuxt is unable to proxy the requests to the backend, once I figure out why I'll be back up and running. 
I really want to make progress on this, thanks for the quick answers lately :)

Edit: Node version 18 has econnrefused error when nuxt proxies requests to the api, switching to node 16 fixes it. Node sux

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 10]
```diff
@@ -0,0 +1,924 @@
...
+type ToolbarIcons<T extends string> = {
+  sectionTitle: string;
+  eventHandler(mode: T): void;
+  highlight: T;
+  icons: {
+    name: T;
+    icon: string;
+    tooltip: string;
+  }[];
+}[];
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: These types should be extracted into an ocr-types file

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+    const toolbarIcons = ref<ToolbarIcons<CanvasModes | SelectedTextSplitModes>>([
+      {
+        sectionTitle: "Toolbar",
+        eventHandler: switchCanvasMode,
+        highlight: state.canvasMode,
+        icons: [
+          {
+            name: "selection",
+            icon: $globals.icons.selectMode,
+            tooltip: i18n.t("ocr-editor.selection-mode") as string,
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: use `tc` method to get a string back instead of casting as string

File: frontend/pages/recipe/_slug/ocr-editor.vue [hunk 12]
```diff
@@ -0,0 +1,924 @@
...
+        tsv.value = res.data as OcrTsvResponse[];
+        state.loading = false;
+
+        if (recipe.value.name?.match(/New\sOCR\sRecipe(\s\([0-9]+\))?/g)) {
+          recipe.value.name = findRecipeTitle();
+        }
+
+        nextTick(() => {
+          state.canvas = <HTMLCanvasElement>document.getElementById("canvas");
+          state.ctx = <CanvasRenderingContext2D>state.canvas.getContext("2d");
```
Comment (hay-kot) [inline frontend/pages/recipe/_slug/ocr-editor.vue]: You should use [template refs](https://vuejs.org/guide/essentials/template-refs.html#template-refs) to access dom elements instead of using document.getElementById.

**Review by hay-kot — COMMENTED** [review]
I tried to go through and clean up what was left to get it merged, but there was too much going on in the Vue component for me to dig though and fix everything. Maybe when I have some more time I can go through it and clean up it, but I left some comments on the issues that I'm seeing

The biggest problem with the component is that there is just so much going on and it's difficult to group what belongs together. What I was going to was break the functions and state that go together into separate composable and place those in a different file to allow for logical group of items - maybe related but there were so many typescript errors in VSCode from Volar that it made reviewing extremely difficult, not sure what the issue was there. 

btw, you've got a `__init.py__` file in the mealie/services/ocr that needs to be fixed.

Like I said, I tried to get it to a good point, but it was more of a weekend project. Will hopefully take another crack at it this weekend if you don't get to it first.

**Comment by Miroito** [comment]
> I tried to go through and clean up what was left to get it merged, but there was too much going on in the Vue component for me to dig though and fix everything. Maybe when I have some more time I can go through it and clean up it, but I left some comments on the issues that I'm seeing
> The biggest problem with the component is that there is just so much going on and it's difficult to group what belongs together. What I was going to was break the functions and state that go together into separate composable and place those in a different file to allow for logical group of items - maybe related but there were so many typescript errors in VSCode from Volar that it made reviewing extremely difficult, not sure what the issue was there.

I can do the clean up if you think it should be done before it is merged. One of the reasons I did not do it yet is that actually, most of the script mess are event handlers for the canvas, which means that if I move the canvas to its own component for example, most of the current functions will follow hence just moving the mess. Most of it is math or helper functions to prevent code duplication. I'll try to make it look as best as I can make it look and you let me know what you think.

Least I can do is give it a try, it will be easier for me to clean up since I wrote all those things.

> btw, you've got a `__init.py__` file in the mealie/services/ocr that needs to be fixed.
Lol yes nice catch, I don't really know what to write inside so I left it empty.


A more general note: Volar is complaining only in the template about the recipe returned by useRecipe because the `Recipe` type has all fields optional. Honestly this is a more general issue that even the new recipe page is not compliant with. The component prop asks for a `NoUndefinedFields<Recipe>` but pages/recipe/_slug/index.vue gives it a `Recipe` anyway. There is just one error reported instead of multiple since the RecipePage component assumes everything is there. This issue alone makes it very difficult to write a proper page using a recipe.
Maybe it would be nicer to have a set of fields that a recipe *must* have, and a set of truly optional properties that we can check on a case by case basis. I mean even the assets (or ingredients, or instructions) property could be an empty array rather than sending back undefined, the existence of undefined and null is pure pain...

**Comment by hay-kot** [comment]
Merged as apart of #1670 with a few minor things cleaned up. Thanks for sticking with this one until the end. 🎉
